Multiple instantiaions of Set cause a C error

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Debian GNU/Linux bookworm/sid
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 5800X 8-Core Processor
CC version: cc (Debian 12.2.0-3) 12.2.0

getwd: /tmp/v_sets
vmodules: /home/mikolaj/.vmodules
vroot: /home/mikolaj/Dokumenty/External-Software/v
vexe: /home/mikolaj/Dokumenty/External-Software/v/v
vexe mtime: 2022-10-14 13:15:13
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.1 6bf2ad1

Git version: git version 2.35.1
Git vroot status: weekly.2022.40-70-g6bf2ad1f
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg foo.v`
```v
import datatypes { Set }

fn numbers() {
	a := [i64(1), 2]
	mut s := Set<i64>{}
	s.add_all(a)

	println(s)
}

fn strings() {
	a := ['1', '2']
	mut s := Set<string>{}
	s.add_all(a)

	println(s)
}

fn main() {
	numbers()
	strings()
}

```

**What did you expect to see?**

No compilation errors (2 different sets printed)


**What did you see instead?**
```
==================
/tmp/v_1000/foo.3055669568369948290.tmp.c:12361: error: invalid operand types for binary operation
/tmp/v_1000/foo.3055669568369948290.tmp.c:12361: error: invalid aggregate type for register load
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1291,10 +1291,10 @@ pub fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 	} else if (left_sym.kind == .map || final_left_sym.kind == .map)
 		&& method_name in ['clone', 'keys', 'values', 'move', 'delete'] {
 		if left_sym.kind == .map {
-			return c.map_builtin_method_call(mut node, left_type, left_sym)
+			return c.map_builtin_method_call(mut node, left_type, c.table.sym(left_type))
 		} else {
 			parent_type := (left_sym.info as ast.Alias).parent_type
-			return c.map_builtin_method_call(mut node, parent_type, final_left_sym)
+			return c.map_builtin_method_call(mut node, parent_type, c.table.final_sym(left_type))
 		}
 	} else if left_sym.kind == .array && method_name in ['insert', 'prepend'] {
 		if method_name == 'insert' {
@@ -2085,6 +2085,12 @@ fn (mut c Checker) map_builtin_method_call(mut node ast.CallExpr, left_type ast.
 				c.table.find_or_register_array(info.value_type)
 			}
 			ret_type = ast.Type(typ)
+			if method_name == 'keys' && info.key_type.has_flag(.generic) {
+				ret_type = ret_type.set_flag(.generic)
+			}
+			if method_name == 'values' && info.value_type.has_flag(.generic) {
+				ret_type = ret_type.set_flag(.generic)
+			}
 		}
 		'delete' {
 			c.fail_if_immutable(node.left)

--- a/vlib/v/tests/map_builtin_call_test.v
+++ b/vlib/v/tests/map_builtin_call_test.v
@@ -8,8 +8,20 @@ fn call_value_is_generic<T>(v T) {
 	_ := a.values().filter(it == v)
 }
 
+fn call_all_is_generic_keys_method<T, U>(v T) {
+	a := map[T]U{}
+	_ := a.keys().filter(it == v)
+}
+
+fn call_all_is_generic_values_method<T, U>(v U) {
+	a := map[T]U{}
+	_ := a.values().filter(it == v)
+}
+
 fn test_call_has_generic() {
 	call_key_is_generic<int>(1)
 	call_value_is_generic<string>('')
+	call_all_is_generic_keys_method<int, string>(1)
+	call_all_is_generic_values_method<int, string>('')
 	assert true
 }

--- a/vlib/v/tests/map_builtin_call_test.v
+++ b/vlib/v/tests/map_builtin_call_test.v
@@ -1,0 +1,15 @@
+fn call_key_is_generic<T>(v T) {
+	a := map[T]u8{}
+	_ := a.keys().filter(it == v)
+}
+
+fn call_value_is_generic<T>(v T) {
+	a := map[u8]T{}
+	_ := a.values().filter(it == v)
+}
+
+fn test_call_has_generic() {
+	call_key_is_generic<int>(1)
+	call_value_is_generic<string>('')
+	assert true
+}

